NWNXLib/Services/PerObjectStorage/PerObjectStorage.cpp

```diff
@@ -0,0 +1,224 @@
+{
+    // TODO cleanup all storage from this plugin
+}
+
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, int value)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value);
+}
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, float value)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value);
+}
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, std::string value)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value);
+}
+void PerObjectStorageProxy::Set(API::Types::ObjectID object, std::string key, void *value, PerObjectStorage::CleanupFunc cleanup)
+{
+    m_proxyBase.Set(object, m_pluginName + "!" + key, value, cleanup);
+}
+
+void PerObjectStorageProxy::Remove(API::Types::ObjectID object, std::string key)
+{
+    m_proxyBase.Remove(object, m_pluginName + "!" + key);
+}
+
+
+template <> Maybe<int> PerObjectStorage::Get<int>(API::Types::ObjectID object, std::string key)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+    {
+        auto map = it->second->GetIntMap();
+        auto it2 = map.find(key);
+        if (it2 != map.end())
+            return Maybe<int>(std::move(it2->second));
+    }
+    return Maybe<int>();
+}
+template <> Maybe<float> PerObjectStorage::Get<float>(API::Types::ObjectID object, std::string key)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+    {
+        auto map = it->second->GetFloatMap();
+        auto it2 = map.find(key);
+        if (it2 != map.end())
+            return Maybe<float>(std::move(it2->second));
+    }
+    return Maybe<float>();
+}
+template <> Maybe<std::string> PerObjectStorage::Get<std::string>(API::Types::ObjectID object, std::string key)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+    {
+        auto map = it->second->GetStringMap();
+        auto it2 = map.find(key);
+        if (it2 != map.end())
+            return Maybe<std::string>(std::move(it2->second));
+    }
+    return Maybe<std::string>();
+}
+
+template <> Maybe<void*> PerObjectStorage::Get<void*>(API::Types::ObjectID object, std::string key)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+    {
+        auto map = it->second->GetPointerMap();
+        auto it2 = map.find(key);
+        if (it2 != map.end())
+            return Maybe<void*>(std::move(it2->second.first));
+    }
+    return Maybe<void*>();
+}
+
+void PerObjectStorage::CGameObjectArray__Delete__1_hook(Services::Hooks::CallType type, API::CGameObjectArray* thisPtr, uint32_t id, API::CGameObject **ptr)
+{
+    // unreferenced variables
+    (void)(sizeof(thisPtr) & sizeof(ptr));
+
+    if (type != Services::Hooks::CallType::AFTER_ORIGINAL)
+        return;
+
+    id &= 0x7FFFFFFF;
+    LOG_DEBUG("Destroying object storage for objectId:0x%08x", id);
```

Comment:
I guess practically speaking we don't want this as it will probably be very very slow ... ?

Reply:
We definitely don't want the printing. I was hoping the perf of an early GTFO is negligible.